compiler/rustc_resolve/src/imports.rs

```diff
@@ -1368,12 +1367,24 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
             let mut redundant_spans: Vec<_> = redundant_span.present_items().collect();
             redundant_spans.sort();
             redundant_spans.dedup();
+            let is_preload = self.extern_prelude_get(target, false).is_some();
```

Comment:
I don't think this is a correct approach, it regresses the 2015 edition case in particular, when we can indeed emit the "the item `{ident}` is already {introduced} here" with a good span (the `extern crate` item).

(Also `ns` is not checked, so `is_preload` can return true for values and macros, and "preload" is not a terminology used anywhere else.)

Reply:
https://github.com/rust-lang/rust/pull/121958#discussion_r1511402323

Reply:
Hi, I am also paying attention to this issue. This detection method may can't cover the situation for https://github.com/rust-lang/rust/issues/121315.

It would be great if both issues could be fixed in this PR.